persistence/src/main/java/com/epam/esm/repository/impl/GiftCertificateDaoImpl.java

```diff
@@ -0,0 +1,165 @@
+        return orders.isEmpty() || orders == null ? false : true;
+    }
+
+    @Override
+    public Long rowsInTable() {
+        CriteriaBuilder cb = entityManager.getCriteriaBuilder();
+        CriteriaQuery<Long> cQuery = cb.createQuery(Long.class);
+        cQuery.select(cb.count(cQuery.from(GiftCertificate.class)));
+        return entityManager.createQuery(cQuery).getSingleResult();
+    }
+
+    @Override
+    public Long rowsInTable(GiftCertificateSearchParams searchParams) {
+        CriteriaBuilder cb = entityManager.getCriteriaBuilder();
+        CriteriaQuery<Long> cQuery = cb.createQuery(Long.class);
+        Root<GiftCertificate> root = cQuery.from(GiftCertificate.class);
+
+        List<Predicate> searchByParameters = setSearchParameters(searchParams, cb, root);
+
+        cQuery.where(cb.and(searchByParameters.toArray(new Predicate[searchByParameters.size()])));
+        cQuery.select(cb.count(root)).where(cb.and(searchByParameters.toArray(new Predicate[searchByParameters.size()])));
+
+        return entityManager.createQuery(cQuery).getSingleResult();
+    }
+
+    private List<Predicate> setSearchParameters(GiftCertificateSearchParams searchParams,
+                                                CriteriaBuilder cb,
+                                                Root<GiftCertificate> root) {
+
+        List<Predicate> searchByParameters = new ArrayList<>();
+
+        if (searchParams.getGiftCertificateName() != null && !searchParams.getGiftCertificateName().isEmpty()) {
+            searchByParameters.add(cb.like(
+                    root.get(GIFT_CERTIFICATE_NAME), "%" + searchParams.getGiftCertificateName() + "%"));
+        }
+        if (searchParams.getGiftCertificateDescription() != null
+                && !searchParams.getGiftCertificateDescription().isEmpty()) {
+            searchByParameters.add(cb.like(
+                    root.get(GIFT_CERTIFICATE_DESCRIPTION), "%" + searchParams.getGiftCertificateDescription() + "%")
+            );
+        }
+
+        if (searchParams.getTags() != null && !searchParams.getTags().isEmpty()) {
+            List<String> tags = searchParams.getTags();
+            for (String tag : tags) {
+                searchByParameters.add(cb.equal(root.join(GIFT_CERTIFICATE_TAGS).get(TAG_NAME), tag));
+            }
+        }
+        return searchByParameters;
+    }
+
+    private void setOrderParameters(CriteriaBuilder cb,
+                                    CriteriaQuery<GiftCertificate> cQuery,
+                                    Root<GiftCertificate> root,
+                                    GiftCertificateSearchParams searchParams) {
+        if (searchParams.getSort() != null && searchParams.getSort().size() > 0) {
+            Set<String> sort = searchParams.getSort();
+            for (String s : sort) {
+                switch (s) {
+                    case "sortByGiftCertificateName-":
```

Comment:
please parse the field name first and save to variable

then parse an ask/desc